Ensure LocalEndPoint being updated in SendToAsync(SocketAsyncEventArgs)

antonfirsov opened the pull request:
Fixes #915 by assigning `_rightEndPoint` like in other overloads, introducing a breaking change in the behavior of `SendToAsync(SocketAsyncEventArgs)`.

Fixes #915

Commit:
tests for implicit binding of LocalEndpoint

Commit:
assign _rightEndPoint in SendToAsync(e)

Fix #26882

Commit:
validate LocalEndPoint asynchronously in tests

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=808) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=808) before we can accept your contribution.<br/><br/>:x: antonfirsov [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=808)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=808) it.</sub>

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -4121,6 +4121,12 @@ public bool SendToAsync(SocketAsyncEventArgs e)
+            if (_rightEndPoint == null)
+            {
+                _rightEndPoint = endPointSnapshot;
+            }
```
I was not sure at which point should this happen. `BeginSendTo` assigns it before doing the actual WSA P/Invoke, I tried to follow that approach.

stephentoub replied:
Nit: for future reference, this can now be written as:
```C#
_rightEndPoint ??= endPointSnapshot;
```

antonfirsov commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/SocketTestHelper.cs:
```diff
@@ -290,6 +290,7 @@ public class SocketHelperSpanSync : SocketHelperArraySync
             Task.Run(() => s.Receive((Span<byte>)buffer, SocketFlags.None));
         public override Task<int> SendAsync(Socket s, ArraySegment<byte> buffer) =>
             Task.Run(() => s.Send((ReadOnlySpan<byte>)buffer, SocketFlags.None));
+        public override bool UsesSync => true;
```
I guess this was missing by mistake. The property seemed useful for my tests.

Commit:
remove unnecessary condition check in tests

Commit:
missing line break

antonfirsov commented:
@scalablecory @davidsh I just realized that my changes are repeating the mistake pointed out in dotnet/corefx#40952. My understanding is that simply replacing the conditional block with `Interlocked.CompareExchange` would fix the race. Is this correct?

Can I address both issues in this PR, or shall I open a separate one?

scalablecory commented:
> @scalablecory @davidsh I just realized that my current changes are repeating the mistake pointed out in [dotnet/corefx#40952](https://github.com/dotnet/corefx/issues/40952). My understanding is that simply replacing the conditional block with `Interlocked.CompareExchange` would fix the race. Is this correct?

It's going to need more than a CAS. Here's what I'm seeing (each snip being a context switch)

1) user calls `BeginRecieveFrom`, and `_rightEndPoint` gets set.

https://github.com/dotnet/runtime/blob/eae85a02786fb2c49b72578f2b0305d38f50da69/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3340-L3343

2) user calls `BeginSendTo` and `_rightEndPoint` is already set, so it is left alone.

https://github.com/dotnet/runtime/blob/eae85a02786fb2c49b72578f2b0305d38f50da69/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L2718-L2721

3) `BeginReceiveFrom` now fails and unsets `_rightEndPoint`

https://github.com/dotnet/runtime/blob/eae85a02786fb2c49b72578f2b0305d38f50da69/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs#L3349-L3353

4) `BeginSendTo` succeeds, and `_rightEndPoint` is null, so `LocalEndPoint` will return `null` despite the socket now being implicitly bound.

> 
> Can I address both issues in this PR, or shall I open a separate one?

I think that's fine.

Commit:
Merge branch 'master' into fix-26882

antonfirsov commented:
/azp run runtime-libraries outerloop

antonfirsov commented:
@dotnet/ncl after a further analysis and test of the code, my opinion got even stronger that we should handle the two topics separately. After we decide exactly how we change the way `_rightEndpoint` is being handled, we can change all it's usages in a separate PR.

I removed the _no merge_ label, and marking this PR ready to review as is.

antonfirsov commented:
A WebSocket and an Http outerloop test [failed on linux](https://dev.azure.com/dnceng/public/_build/results?buildId=460101&view=ms.vss-test-web.build-test-results-tab).
I'm not familiar with those codebases yet, but I think it's very unlikely that they utilize `SendToAsync(SocketAsyncEventArgs)`.

scalablecory reviewed (commented):
The other methods null out `_rightEndPoint` if the API call fails. We should be doing this here too, in the `catch` block below your change.

Commit:
Merge remote-tracking branch 'upstream/master' into fix-26882

Commit:
add failing test, refactor tests

wfurt commented:
fixes #915

Commit:
add argument validation tests + clean up test code

Commit:
address review finding (reset _rightEndPoint)

antonfirsov commented on src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.cs:
```diff
@@ -4121,18 +4121,31 @@ public bool SendToAsync(SocketAsyncEventArgs e)
             }
 
+            if (!CheckErrorAndUpdateStatus(socketError))
+            {
```
I have no idea what black magic could trigger the catch block here, on the other hand `socketError == AccessDenied` is something we can actually hit, see my test cases. Like in `DoBeginSendTo` I'm resetting `_rightEndPoint` when hitting a non-success & non-pending case.

antonfirsov commented:
@scalablecory  just addressed your findings + did some test improvements:
- Moved the new `SendTo` tests to it's own file & test classes
- Added missing argument validation tests for all `SendTo` variants

scalablecory commented on src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs:
```diff
@@ -44,6 +44,7 @@ public async Task InvalidArguments_Throws(int? length, int offset, int count)
             }
         }
 
+
```
nit: extra line.

scalablecory reviewed (approved):
lgtm

Commit:
remove extra newline

antonfirsov commented:
Test failures are tracked under #1332 and unrelated to Sockets.

danmosemsft commented:
@antonfirsov is this missing a breaking change doc issue? If so can you please open one with https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md
